concent_api/concent_api/middleware.py

```diff
@@ -39,3 +49,62 @@ def __call__(self, request):
         response = self.get_response(request)
         response['Concent-Version'] = self._concent_version
         return response
+
+
+def determine_return_type(request_meta):
+    try:
+        preffered_types = ['application/json', 'text/html']
+        if "HTTP_ACCEPT" in request_meta:
+            return best_match(preffered_types, request_meta['HTTP_ACCEPT'])
+        return "application/json"
+    except Exception:  # pylint: disable=broad-except
+        return ""
```

Comment:
No. The only thing that can fail is the call to `best_match`. What would you like to fix if it fails? It can fail only if there is something wrong with `HTTP_ACCEPT` and in such case we return `HTTP406` (see `HandleServerErrorMiddleware.process_exception`).

Imagine the processing:
1) Some kind of unhandled internal server error occurs.
2) It is intercepted by the middleware
3) Best return type is being determined
4) `HTTP_ACCEPT` header is somehow broken
5) We return `HTTP406`

In your case:
5") We fail in unhandled error handling middleware.
Is it what we really want? If so, I don't get the point of havin this new middleware.

Reply:
But does `best_match()` raise just plain `Exception`? Doesn't it have any specific error type? You should be catching only that type.

Also, the handler - especially such a broad handler - should have the smallest scope possible to avoid accidentally catching other errors. So if the only thing you think is going to fail is `best_match()` the handler should cover it and nothing else. Not even the `if` around it.

This handler might make sense if you just logged the error (including sending a report to Sentry) and re-raised immediately after because a crash here will not go through the usual error handling Django does for a request.

> Is it what we really want? If so, I don't get the point of havin this new middleware.

Yes. Exactly. Well, I see that the real problem here is that we have fundamentally different approaches to error handling. So let me explain mine:

There are two types of errors: failures and bugs.
1. A **failure** is dependent on external conditions and is inherent in the design of the system. For example running out of disk space, connection error due to server being down, timeout, etc. Failure is inevitable. The final product will always have to deal with failures. We want to detect them and handle them gracefully, though most of the time all we can do is to catch the error and display a message. You cannot predict a failure but you can predict the types of failures that the design allows for and be prepared to handle them.
2. A **bug** is a defect in the code or design. It can be corrected. It should be caught as soon as possible and ideally the final product should contain no bugs. That's rarely possible but the number of bugs left is inversely correlated with the amount of testing the product goes through you do so you can at least work towards eliminating them. But for the testing to be effective it must be easy to detect errors. That's why we want the product to fail loudly whenever anything seems out of ordinary. That's also why I'm so keen on `assert` - you can use it to turn some discrepancy that would go unnoticed into a nice crash that can't be ignored.

My approach is simple: every crash is a bug unless proven otherwise. If you get an exception it's either one of a few specific types indicating a failure that you know you have to handle or something unexpected that should be let through.

During development and Q&A the code goes through exhaustive testing and this approach means that the most egregious bugs are quickly identified and fixed.

Yeah, this means that if we don't find a bug in that phase then some unfortunate client might run into it and get a crash instead of a useful response. But while this may not be good for that client, I think it's best for all clients overall. The damage is limited because we're notified about the crash immediately and fix it in the next release. The alternative is that we silence the problem for the sake of that one client even though we don't really know what what's wrong and other clients keep running into it time and time again without us knowing anything about it.

That's why for me `except Exception:` or `except:` are pretty much a bad smell in the code. They immediately stand out in the review. There are rare cases when their use is warranted but most of the time when I see it it's someone trying to handle failures in a way that unnecessarily silences exceptions caused by bugs. That's why pylint even has the `broad-except` warning. If it wasn't overused it would not be warning about it :)

Sorry for stopping this pull request but I'm quite adamant about this approach. If you don't agree we'll have to talk it over.

Reply:
@cameel , well, I agree to a point. Yes, I can admit, we can divide the perfect world to:
1) bugs
2) failures,
but in my opinion there is some middle ground. An example from our world: we are trying to to get Subtask out of DB at point of the flow when it should be there. Biut for whatever reason it is not. Is it a **bug**? Might be. Or maybe some has been tinkering with the DB (and that caused a **failure**)?

My approach is to deliver the best service to the customers. If we cannot do anything sensible with some unexpected situation, we should fail. But if we can, why should we? Even if there is some bug, we have logging and can use it to find the root cause and provide the fix. And the customer doesn't have to be affected.

This discussion should take place in some meeting with the broader audience.

Getting back to this particular situation:
1) As I wrote before, the only entity that can raise exeption is `best_match` function
2) Having this `try... except` block as it is, is IMHO the most readable way I can think of.
3) As `best_match` is third party function, I try to handle all exceptions it may raise.
4) Example with `SyntaxError` doesn't seem to be adequate here - with modern tools and our `lint`, `fullcheck`, etc. script it is **very** unlikely to get it and not realise it.
5) Function, by the design, should be able to determine which return type should be used (in `Java` I would call this function `throw-safe`)

To sum up. If my arguments do not convince you, then this task should remain opened until the more in-depth discussion can take place. Or someone can take it up from here and implement the desired way (what would it look like, btw?). At the end of the day, you are the architect, so you are the oracle.

Reply:
So we'll have to talk about it because I unfortunately cannot fully agree.

I think it's safer to err on the side of crashing too much. From the code I've seen in various projects I think people have a tendency to catch too much and to treat bugs like failures rather than the other way around.

> but in my opinion there is some middle ground. An example from our world: we are trying to to get Subtask out of DB at point of the flow when it should be there. Biut for whatever reason it is not. Is it a bug? Might be. Or maybe some has been tinkering with the DB (and that caused a failure)?

Definitely a bug as long as the database is under the control of our app. I would agree if there was another client indepently accessing the database but in Django apps that's pretty clear cut for me. They're so tightly tied to their database structure that I consider it fully under their control.

There is some middle ground and that's why there are rare cases where a handler for `Exception` is appropriate. But they are really rare. Like, there might be one or two cases in our whole application (probably in error handling decorators).

> My approach is to deliver the best service to the customers. If we cannot do anything sensible with some unexpected situation, we should fail. But if we can, why should we? 

Because if you catch `Exception` you don't know what you're dealing with. In many situations the safest thing to do is actually to fail and terminate rather than go on and do something wrong. That's my approach - try to make the best of the situation when you can. But don't try to guess when you don't have a clue what's happening.

Simply, it's fine to catch specific exceptions from `best_match()`. Not `Exception`.